File: tasks (demian)/block 4/day 19/task 1/task 1.py
mat = []
for i in range(10):
    mat.append(list(map(int, input().split())))

first = list(map(int, input().split()))
second = list(map(int, input().split()))

test = int(input())
while True:
    if test == 0:
        break

    a, b = map(str, input().split())
    answer = []
    _a = ["."] * 1000
    _b = ["."] * 1000

    maxlen = max(len(a), len(b))

    for i in range(len(str(a))):
        _a[i] = int(a[::-1][i])
    for i in range(len(str(b))):
        _b[i] = int(b[::-1][i])

    _a = _a[:maxlen]
    _b = _b[:maxlen]

    answer = [0] * (maxlen + 1)

    for i in range(maxlen):
        if _a[i] != "." and _b[i] != ".":
            answer[i] = mat[_a[i]][_b[i]]
        elif _a[i] != ".":
            answer[i] = second[_a[i]]
        elif _b[i] != ".":
            answer[i] = first[_b[i]]

    print(*answer[::-1])

    test -= 1

File: OPT/tinkoff generation/day 6 (segment tree)/task 1.cpp
const int lg = 18, maxn = 1e5 + 1;
vector<int> a(maxn), deg(maxn);
vector<vector<pair<int, int>>> sparse(lg, vector<pair<int, int>>(maxn));

pair<int, int> get(int l, int r) {
    int k = deg[r - l + 1];
    pair<int, int> answer;
    answer.first = max(sparse[k][l].first, sparse[k][r - (1 << k) + 1].first);
    answer.second = (sparse[k][l].first > sparse[k][r - (1 << k) + 1].first
                         ? sparse[k][l].second
                         : sparse[k][r - (1 << k) + 1].second);
    return answer;
}

int n;
cin >> n;
for (int i = 1; i <= n; i++) cin >> a[i];
for (int i = 1; i <= n; ++i) {
    sparse[0][i].first = a[i];
    sparse[0][i].second = i;
}

for (int k = 0; k < lg - 1; ++k)
    for (int i = 1; i <= n; ++i) {
        int j = min(i + (1 << k), n);
        sparse[k + 1][i].first = max(sparse[k][i].first, sparse[k][j].first);
        sparse[k + 1][i].second = (sparse[k + 1][i].first == sparse[k][i].first
                                       ? sparse[k][i].second
                                       : sparse[k][j].second);
    }

deg[1] = 0;
for (int a = 2; a <= n; ++a) {
    deg[a] = deg[a - 1];
    if (!(a & (a - 1))) ++deg[a];
}

int k, l, r;
cin >> k;
while (k--) {
    cin >> l >> r;
    pair<int, int> ans = get(l, r);
    cout << ans.first << ' ' << ans.second << '\n';
}